---
title: "Single Cell Analysis"
author: "Max Mattessich, Joaquin Reyna, Anna Konstorum, Edel Aron"
date: "Compiled: `r format(Sys.time(), '%d %B %Y')`"
header-includes:
   - \usepackage{amsmath}
   - \usepackage{amsfonts}
output: 
    rmdformats::readthedown
vignette: >
  %\VignetteIndexEntry{Single-Cell-Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup

You will need the [TENxPBMCData](https://bioconductor.org/packages/3.15/data/experiment/html/TENxPBMCData.html) package from Bioconductor. It contains 9 different publicly available datasets from 10x Genomics.

```{r load-packages, echo = FALSE, include = FALSE}
library(ggplot2)
library(nipalsMCIA)
library(TENxPBMCData)
```

## Loading the data

We will be using the "pbmc5k-CITEseq" dataset, which was published in May 2019. It contains 5,247 detected cells from PBMCSs and was processed using Cell Ranger 3.0.2

### Bioconductor

```{r load-data, eval = FALSE}
tenx_pbmc5k_CITEseq <- TENxPBMCData(dataset = "pbmc5k-CITEseq")

# RNA-Seq data
counts(tenx_pbmc5k_CITEseq)

# CITE-Seq data
counts(altExp(tenx_pbmc5k_CITEseq))
```

```{r}
# tenx_pbmc5k <- data(data_blocks_sc)
load(file.path("..", "data", "data_blocks_sc.Rda"))
```


### From 10x Genomics

*Or other sources...*


### Seurat

```{r, eval = FALSE}
library(Seurat)
tenx_pbmc5k <- as.Seurat(tenx_pbmc5k_CITEseq, counts = "counts", data = "logcounts")
```

## scRNA-Seq analysis

### Read in the data

```{r}
# load the data
data <- Read10X(file.path("..", "data", "tenx_pbmc5k_CITEseq", "filtered_feature_bc_matrix"))

# create the Seurat object
obj <- CreateSeuratObject(counts = data$`Gene Expression`)
adt_assay <- CreateAssayObject(counts = data$`Antibody Capture`)
obj[["ADT"]] <- adt_assay

# add useful metadata
obj$"percent.mt" <- PercentageFeatureSet(obj, pattern = "^MT-")

# reformat the barcodes
# obj <- RenameCells(obj, new.names = substr(Cells(obj), 1, 16)) # remove the -1s

# save the object
saveRDS(obj, file.path("..", "data", "tenx_pbmc5k_CITEseq_unprocessed.rds"))
```


### Quality control


#### Metrics summary

```{r}
library(readr)

# read in the summary table
metrics_summary <- read_csv(file.path("..", "data", "tenx_pbmc5k_CITEseq", 
                                      "5k_pbmc_protein_v3_metrics_summary.csv"),
                            show_col_types = FALSE)
```


#### GEX QC metrics

##### Before filtering

```{r}
VlnPlot(obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), 
        pt.size = 0.01, ncol = 3)
```

##### After filtering

```{r}
obj <- subset(obj, subset = nFeature_RNA > 200 & percent.mt < 20)

VlnPlot(obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), 
        pt.size = 0.01, ncol = 3)
```
### Standard Seurat pipeline

```{r}
# standard normalization
obj <- NormalizeData(obj, normalization.method = "LogNormalize", scale.factor = 10000)

# highly variable features
obj <- FindVariableFeatures(obj, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(obj), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(obj)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2

# scaling
obj <- ScaleData(obj, features = rownames(obj))

# dimensionality reduction
obj <- RunPCA(obj) # verbose = FALSE

ElbowPlot(obj)

# clustering
obj <- FindNeighbors(obj, reduction = "pca", dims = 1:20) # verbose = FALSE
obj <- FindClusters(obj, resolution = 0.4) # verbose = FALSE
obj <- RunUMAP(obj, reduction = "pca", dims = 1:20) # verbose = FALSE

# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
DimPlot(obj, reduction = "umap")
```


## Match NCI60 structure

*Just for testing*

```{r examine-structure}
# import data as "data_blocks" and metadata as "metadata_NCI60"
data(NCI60)

summary(data_blocks)

metadata_NCI60

# data_blocks$mrna
```

```{r match-structure}
data_blocks_sc <- list()
data_blocks_sc$mrna <- data.frame(as.matrix(counts(tenx_pbmc5k_CITEseq)))
data_blocks_sc$adt <- data.frame(as.matrix(counts(altExp(tenx_pbmc5k_CITEseq))))

summary(data_blocks_sc)

save(data_blocks_sc, file = "data_blocks_sc.Rda")
```


## Results

```{r original}
mcia_results <- nipals_multiblock(data_blocks, preprocMethod = 'colprofile',
                                  metadata = metadata_NCI60, coloring = "cancerType", 
                                  num_PCs = 10, tol = 1e-12)
```


```{r sc, eval = FALSE}
mcia_results_sc <- nipals_multiblock(data_blocks_sc, preprocMethod = 'colprofile',
                                  # metadata = metadata_NCI60, coloring = "cancerType", 
                                  num_PCs = 5, tol = 1e-12)
```


## Tests

From Anna

```{r}
data_blocks_sc_mod <- readRDS(file.path("..", "data", "data_blocks_sc_mod.RDS"))
data_blocks_sc_mod$mrna <- t(data_blocks_sc_mod$mrna)
data_blocks_sc_mod$adt <- t(data_blocks_sc_mod$adt)
mcia_results <- nipals_multiblock(data_blocks_sc_mod, num_PCs = 2, tol = 1e-12,preprocMethod="none")

saveRDS(mcia_results,"mcia_results_sc.RDS")
MCIA_plots(mcia_results, "projection_global", orders = c(1,2), legend_loc = "bottomleft")
```

