---
title: "Single Cell Analysis"
author: "Max Mattessich, Joaquin Reyna, Edel Aron, Anna Konstorum"
date: "Compiled: `r format(Sys.time(), '%d %B %Y')`"
header-includes:
   - \usepackage{amsmath}
   - \usepackage{amsfonts}
output: 
    rmdformats::readthedown:
      toc_depth: 4
vignette: >
  %\VignetteIndexEntry{Single-Cell-Analysis}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
set.seed(42) # NIPALS starts with a random vector
```

## Setup

Note that the `TENxPBMCData` package is not included in this list as you may decide to pull data from another source.

```{r load-packages, echo = FALSE, include = FALSE}
library(dplyr)
library(ggplot2)
library(nipalsMCIA)
library(piggyback)
library(Seurat)
```

## Loading the data

We will be using the 10x Genomics "pbmc5k-CITEseq" dataset, which was published in May 2019 and processed using Cell Ranger 3.0.2. It contains 5,247 detected cells from PBMCs and 33,538 genes. We chose it due to it containing both GEX and ADT data as well as being relatively recent and from a widely used company.

The following detail several ways in which to load in this dataset. You may prefer to use data from other sources outside of 10x Genomics, in which case you will have to format it to work with nipalsMCIA.

### Bioconductor

There are several Bioconductor packages which provide single-cell data for users as part of the package. The [TENxPBMCData](https://bioconductor.org/packages/3.15/data/experiment/html/TENxPBMCData.html) contains 9 different publicly available datasets from 10x Genomics, including the one that we will be analyzing.

```{r, eval = FALSE}
# if needed
# install.packages(TENxPBMCData)

# read in the data as a SingleCellExperiment object
data_blocks_sc <- TENxPBMCData::TENxPBMCData(dataset = "pbmc5k-CITEseq")

# examine the data
counts(data_blocks_sc) # RNA-Seq data ("mainExpName: Gene Expression")
counts(altExp(data_blocks_sc)) # CITE-Seq data ("altExpNames(1): Antibody Capture")
```
In order to run...

```{r, include = FALSE, eval = FALSE}
# testing methods
tenx_pbmc3k <- TENxPBMCData(dataset = "pbmc3k")
data_blocks_sc_test <- data_blocks_sc

data_blocks_sc <- as.Seurat(data_blocks_sc, counts = "counts", data = "logcounts")

library(Matrix)
pbmc.data <- as(counts(data_blocks_sc), "Matrix")
pbmc.data <- as(pbmc.data, "dgTMatrix")
colnames(pbmc.data) <- paste0("Cell", seq_len(ncol(pbmc.data)))
rownames(pbmc.data) <- make.unique(rowData(data_blocks_sc)[, "Symbol_TENx", drop=TRUE])
pbmc <- CreateSeuratObject(raw.data = pbmc.data, min.cells = 3, min.genes = 200, project = "10X_PBMC")

rownames(data_blocks_sc_test) <- scater::uniquifyFeatureNames(rowData(data_blocks_sc_test)$ENSEMBL_ID,
                                                      rowData(data_blocks_sc_test)$Symbol_TENx)

pbmc <- as.Seurat(data_blocks_sc_test, data = NULL)
```

### 10x Genomics

The original dataset can be found on the [10x Genomics Datasets website](https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.2/5k_pbmc_protein_v3) and an explanation of the file types for this version can be found [here](https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/3.0/output/overview). You can download them all to a directory of your choosing with their suggested commands:

```{bash, eval = FALSE}
# Input Files
curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_protein_v3/5k_pbmc_protein_v3_fastqs.tar
curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_protein_v3/5k_pbmc_protein_v3_feature_ref.csv

# Output Files
curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_protein_v3/5k_pbmc_protein_v3_possorted_genome_bam.bam
curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_protein_v3/5k_pbmc_protein_v3_possorted_genome_bam.bam.bai
curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_protein_v3/5k_pbmc_protein_v3_molecule_info.h5
curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_protein_v3/5k_pbmc_protein_v3_filtered_feature_bc_matrix.h5
curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_protein_v3/5k_pbmc_protein_v3_filtered_feature_bc_matrix.tar.gz
curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_protein_v3/5k_pbmc_protein_v3_raw_feature_bc_matrix.h5
curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_protein_v3/5k_pbmc_protein_v3_raw_feature_bc_matrix.tar.gz
curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_protein_v3/5k_pbmc_protein_v3_analysis.tar.gz
curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_protein_v3/5k_pbmc_protein_v3_metrics_summary.csv
curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_protein_v3/5k_pbmc_protein_v3_web_summary.html
curl -O https://cf.10xgenomics.com/samples/cell-exp/3.0.2/5k_pbmc_protein_v3/5k_pbmc_protein_v3_cloupe.cloupe
```

You will need the *filtered_feature_bc_matrix* directory for the following analysis; it can be extracted with `tar -xvzf 5k_pbmc_protein_v3_filtered_feature_bc_matrix.tar.gz` from within the relevant data directory.

```{r, include = TRUE, eval = FALSE}
# load the data (change the file path as needed)
data <- Read10X(file.path("..", "data", "tenx_pbmc5k_CITEseq", "filtered_feature_bc_matrix"))
```

### nipalsMCIA

```{r, include = TRUE, eval = FALSE}
# download the data from the repository if needed
pb_download(file = "data_blocks_sc.Rda", repo = "Muunraker/nipalsMCIA", 
            tag = "v0.0.1", dest = file.path("..", "data"))

# data(data_blocks_sc)
```

```{r, include = TRUE, eval = FALSE}
# load the data
# data_blocks_sc <- load(file = file.path("..", "data", "data_blocks_sc.Rda"))
data_blocks_sc <- load(file = "../data/data_blocks_sc.Rda")

# examine the data
summary(data_blocks_sc)

# metadata_NCI60
```

### Seurat

```{r}

```


## scRNA-Seq analysis

### Read in the data

```{r, include = TRUE, message = FALSE}
# load the data
data <- Read10X(file.path("..", "data", "tenx_pbmc5k_CITEseq", "filtered_feature_bc_matrix"))

# create the Seurat object
obj <- CreateSeuratObject(counts = data$`Gene Expression`)
adt_assay <- CreateAssayObject(counts = data$`Antibody Capture`)
obj[["ADT"]] <- adt_assay

# add useful metadata
obj$"percent.mt" <- PercentageFeatureSet(obj, pattern = "^MT-")

# reformat the barcodes
# obj <- RenameCells(obj, new.names = substr(Cells(obj), 1, 16)) # remove the -1s

# save the object
# saveRDS(obj, file.path("..", "data", "tenx_pbmc5k_CITEseq_unprocessed.rds"))
```


### Quality control


#### Metrics summary

```{r, include = TRUE}
library(readr)

# read in the summary table
metrics_summary <- read_csv(file.path("..", "data", "tenx_pbmc5k_CITEseq", 
                                      "5k_pbmc_protein_v3_metrics_summary.csv"),
                            show_col_types = FALSE)
```


#### GEX QC metrics

##### Before filtering

```{r, include = TRUE}
VlnPlot(obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), 
        pt.size = 0.01, ncol = 3)
```

##### After filtering

```{r, include = TRUE}
obj <- subset(obj, subset = nFeature_RNA > 200 & percent.mt < 20)

VlnPlot(obj, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), 
        pt.size = 0.01, ncol = 3)
```

### Standard Seurat pipeline

```{r, include = TRUE, eval = FALSE}
# standard normalization
obj <- NormalizeData(obj, normalization.method = "LogNormalize", scale.factor = 10000)

# highly variable features
obj <- FindVariableFeatures(obj, selection.method = "vst", nfeatures = 2000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(obj), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(obj)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot1 + plot2

# scaling
obj <- ScaleData(obj, features = rownames(obj))

# dimensionality reduction
obj <- RunPCA(obj) # verbose = FALSE

ElbowPlot(obj)

# clustering
obj <- FindNeighbors(obj, reduction = "pca", dims = 1:20) # verbose = FALSE
obj <- FindClusters(obj, resolution = 0.4) # verbose = FALSE
obj <- RunUMAP(obj, reduction = "pca", dims = 1:20) # verbose = FALSE

# saveRDS(obj, file.path("..", "data", "tenx_pbmc5k_CITEseq_processed.rds"))
```

```{r, include = TRUE}
obj <- readRDS(file.path("..", "data", "tenx_pbmc5k_CITEseq_processed.rds"))

# note that you can set `label = TRUE` or use the LabelClusters function to help label
# individual clusters
plot_seurat_clusters <- UMAPPlot(obj)
plot_seurat_clusters
```

### Marker overlays

#### Load marker genes

```{r, include = TRUE}
# sort by Cell_Type and Marker if not already sorted
markers_all <- read_csv(file.path("..", "data", "marker_genes.csv"),
                        show_col_types = FALSE)
markers_all
```


#### Dot plots

```{r, include = TRUE}
p <- DotPlot(obj, features = unique(markers_all$Marker),
             dot.scale = 3, cluster.idents = TRUE) + RotatedAxis()

p$data <- suppressMessages(left_join(p$data, markers_all %>%
                                             rename(features.plot = Marker)))

p + facet_grid(cols = vars(Cell_Type), scales = 'free_x', space = "free") +
      theme(strip.text.x = element_text(size = 8))
```


#### Feature plots

```{r, include = TRUE}
FeaturePlot(obj, features = unique(markers_all$Marker), label = TRUE)
```

### Annotate cell clusters

```{r, include = TRUE}
obj_annotations <- rbind(c("0", "Unknown"),
                         c("1", "T cells"),
                         c("2", "T cells"),
                         c("3", "Unknown"),
                         c("4", "Natural killers"),
                         c("5", "B cells"),
                         c("6", "Unknown"),
                         c("7", "Unknown"),
                         c("8", "Unknown"),
                         c("9", "pDCs"))
colnames(obj_annotations) <- c("Cluster", "CellType")
obj_annotations <- data.frame(obj_annotations)

# prepare the annotation information
annotations <- obj_annotations[["CellType"]] # you only need the cell type information
names(annotations) <- levels(obj$seurat_clusters)

# relabel the Seurat clusters
# Idents(obj) <- "seurat_clusters"
obj <- RenameIdents(obj, annotations)

# alphabetize the cell types
Idents(obj) <- factor(Idents(obj), levels = sort(levels(obj)))

# useful metadata (e.g. if you want to have multiple annotation sets)
obj[["annotated_clusters"]] <- Idents(obj)

# info about the clusters
obj_annotations %>%
  group_by(CellType) %>%
  transmute(Clusters = paste0(Cluster, collapse = ", ")) %>%
  distinct() %>% arrange(CellType)
```

```{r, include = TRUE}
plot_annotated_clusters <- UMAPPlot(object = obj, label = TRUE)

plot_seurat_clusters + plot_annotated_clusters
```


## Save the data

```{r match-structure, include = TRUE, eval = FALSE}
data_blocks_sc <- list()
data_blocks_sc$mrna <- data.frame(as.matrix(counts(tenx_pbmc5k_CITEseq)))
data_blocks_sc$adt <- data.frame(as.matrix(counts(altExp(tenx_pbmc5k_CITEseq))))

summary(data_blocks_sc)

save(data_blocks_sc, file = "data_blocks_sc.Rda")
```

## MCIA

```{r original, include = TRUE, eval = FALSE}
meta_colors = c('orange', 'purple', 'pink')
mcia_results <- nipals_multiblock(data_blocks, preprocMethod = 'colprofile',
                                  metadata = metadata_NCI60, color_col = "cancerType", 
                                  color_pal = meta_colors, 
                                  num_PCs = 10, tol = 1e-12)
```

```{r sc, include = TRUE, eval = FALSE}
mcia_results_sc <- nipals_multiblock(data_blocks_sc, preprocMethod = 'colprofile',
                                     # metadata = metadata_NCI60, color_col = "cancerType", 
                                     num_PCs = 5, tol = 1e-12)
```

## Tests

From Anna

```{r, eval = FALSE}
data_blocks_sc_mod <- readRDS(file.path("..", "data", "data_blocks_sc_mod.RDS"))
data_blocks_sc_mod$mrna <- t(data_blocks_sc_mod$mrna)
data_blocks_sc_mod$adt <- t(data_blocks_sc_mod$adt)
mcia_results <- nipals_multiblock(data_blocks_sc_mod, num_PCs = 2, tol = 1e-12,preprocMethod="none")

saveRDS(mcia_results,"mcia_results_sc.RDS")
MCIA_plots(mcia_results, "projection_global", orders = c(1,2), legend_loc = "bottomleft")
```

## Session Info

```{r}
sessionInfo()
```
