---
title: "Single cell analysis"
author: "Edel Aron"
date: '2022-09-20'
header-includes:
   - \usepackage{amsmath}
   - \usepackage{amsfonts}
output:
  html_document: default
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup

You will need the [TENxPBMCData](https://bioconductor.org/packages/3.15/data/experiment/html/TENxPBMCData.html) package from Bioconductor. It contains 9 different publicly available datasets from 10x Genomics.

```{r load-packages, echo = FALSE, include = FALSE}
library(TENxPBMCData)
library(nipalsMCIA)
```


## Loading the data

We will be using the "pbmc5k-CITEseq" dataset, which was published in May 2019. It contains 5,247 detected cells from PBMCSs and was processed using Cell Ranger 3.0.2

```{r load-data}
tenx_pbmc5k_CITEseq <- TENxPBMCData(dataset = "pbmc5k-CITEseq")

# RNA-Seq data
counts(tenx_pbmc5k_CITEseq)

# CITE-Seq data
counts(altExp(tenx_pbmc5k_CITEseq))
```

